### PR TITLE
pkg/profiler: add deregistration for counter metric

### PR DESCRIPTION
**Pull request by Sylfrena**
Signed-off-by: Sumera Priyadarsini <sylphrenadin@gmail.com>

### CODE REVIEW

File: cmd/parca-agent/main.go [hunk 1]
```diff
@@ -159,7 +160,14 @@ func main() {
...
 	externalLabels := getExternalLabels(flags.ExternalLabel, flags.Node)
-	tm := target.NewManager(logger, reg, externalLabels, ksymCache, profileListener, debugInfoClient, flags.ProfilingDuration, flags.TempDir)
+	counter := promauto.With(reg).NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "parca_agent_profiler_missing_stacks_total",
+			Help: "Number of missing profile stacks",
+		},
+		[]string{"type"})
+
+	tm := target.NewManager(logger, counter, externalLabels, ksymCache, profileListener, debugInfoClient, flags.ProfilingDuration, flags.TempDir)
```
Comment (Sylfrena) [inline cmd/parca-agent/main.go:171]: Create a metrics struct in `profiler.go` and add a helper function to initialise it maybe.

File: pkg/profiler/profiler.go [hunk 2]
```diff
@@ -132,6 +125,10 @@ func (p *CgroupProfiler) Stop() {
 	p.mtx.Lock()
 	defer p.mtx.Unlock()
 	level.Debug(p.logger).Log("msg", "stopping cgroup profiler")
+	err := p.missingStacks.DeleteLabelValues()
+	if !err {
+		fmt.Errorf("error deleting profiles")
```
Comment (Sylfrena) [inline pkg/profiler/profiler.go:130]: also, like, find a better way to do this.

Comment (metalmatze) [reply pkg/profiler/profiler.go:130]: What do you mean with better? This should be totally fine to do :slightly_smiling_face: 
```suggestion
		return fmt.Errorf("error deleting profiles: %w", err)
```

File: pkg/profiler/profiler.go [hunk 3]
```diff
@@ -132,6 +125,10 @@ func (p *CgroupProfiler) Stop() {
 	p.mtx.Lock()
 	defer p.mtx.Unlock()
 	level.Debug(p.logger).Log("msg", "stopping cgroup profiler")
+	err := p.missingStacks.DeleteLabelValues()
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply pkg/profiler/profiler.go:128]: I quickly tried it

https://go.dev/play/p/ozETX_wVQtt

**Review by kakkoyun — CHANGES_REQUESTED** [review]
I have left some comments.

File: pkg/profiler/profiler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kakkoyun) [reply pkg/profiler/profiler.go:128]: @brancz WDYT?

File: cmd/parca-agent/main.go [hunk 4]
```diff
@@ -159,7 +160,14 @@ func main() {
 	}
 
 	externalLabels := getExternalLabels(flags.ExternalLabel, flags.Node)
-	tm := target.NewManager(logger, reg, externalLabels, ksymCache, profileListener, debugInfoClient, flags.ProfilingDuration, flags.TempDir)
+	counter := promauto.With(reg).NewCounterVec(
```
[the remark this answers is not in the document]
Comment (Sylfrena) [reply cmd/parca-agent/main.go:163]: Yep, you are right, will make the changes.

File: pkg/profiler/profiler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (brancz) [reply pkg/profiler/profiler.go:128]: I get a build error with the go playground example, but it's not important I see what you meant.

I think either way is a bit unnatural. We have a bit of a dilemma, either a counter vec with a const label that is not really constant or a leaky abstraction. I think after seeing your approach I'm also leaning more towards the first.

**Comment by Sylfrena** [comment]
@brancz @kakkoyun,  PTAL

**Review by kakkoyun — APPROVED** [review]
If it solves all the problem, it looks amazing to me :100: 
I love one-liners :stuck_out_tongue_winking_eye:

File: pkg/profiler/profiler.go [hunk 5]
```diff
@@ -132,6 +132,9 @@ func (p *CgroupProfiler) Stop() {
 	p.mtx.Lock()
 	defer p.mtx.Unlock()
 	level.Debug(p.logger).Log("msg", "stopping cgroup profiler")
+	if !prometheus.Unregister(p.missingStacks) {
```
Comment (kakkoyun) [inline pkg/profiler/profiler.go:135]: I just realized after merge, this is not exactly correct.
This unregisters from the default registry. It suppose to be register that registered the metric so the one in the profiler.

**Pull request merged**